File: backend/apps/insights/schemas.py
from datetime import datetime
from typing import Any
from uuid import UUID

from ninja import Schema
from pydantic import Field


class CategoryCreateSchema(Schema):
    name: str = Field(min_length=1, max_length=150)
    slug: str = Field(min_length=1, max_length=170)
    description: str = ""
    parent_id: UUID | None = None
    is_active: bool = True
    sort_order: int = Field(default=0, ge=0)


class CategorySchema(Schema):
    id: UUID
    name: str
    slug: str
    description: str
    parent_id: UUID | None
    is_active: bool
    sort_order: int
    created_at: datetime
    updated_at: datetime


class TagCreateSchema(Schema):
    name: str = Field(min_length=1, max_length=100)
    slug: str = Field(min_length=1, max_length=120)
    description: str = ""
    is_active: bool = True


class TagSchema(Schema):
    id: UUID
    name: str
    slug: str
    description: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


class InternalLinkInputSchema(Schema):
    target_article_id: UUID
    anchor_text: str = Field(min_length=1, max_length=250)
    context: str = ""
    is_active: bool = True


class InsightSeoInputSchema(Schema):
    meta_title: str = Field(default="", max_length=70)
    meta_description: str = Field(
        default="",
        max_length=170,
    )
    canonical_url: str = ""

    robots_index: bool = True
    robots_follow: bool = True

    open_graph_title: str = Field(
        default="",
        max_length=100,
    )
    open_graph_description: str = Field(
        default="",
        max_length=200,
    )
    open_graph_image_id: UUID | None = None

    twitter_title: str = Field(
        default="",
        max_length=100,
    )
    twitter_description: str = Field(
        default="",
        max_length=200,
    )

    article_schema: dict[str, Any] = Field(
        default_factory=dict,
    )
    faq_schema: list[dict[str, Any]] = Field(
        default_factory=list,
    )


class InsightArticleCreateSchema(Schema):
    title: str = Field(min_length=1, max_length=300)
    slug: str = Field(min_length=1, max_length=320)
    excerpt: str = ""
    content: dict[str, Any] = Field(default_factory=dict)

    category_id: UUID | None = None
    tag_ids: list[UUID] = Field(default_factory=list)
    author_id: int | None = None
    featured_image_id: UUID | None = None

    status: str = "draft"
    is_featured: bool = False
    is_active: bool = True
    allow_comments: bool = False

    related_article_ids: list[UUID] = Field(
        default_factory=list,
    )

    internal_links: list[
        InternalLinkInputSchema
    ] = Field(default_factory=list)

    seo: InsightSeoInputSchema = Field(
        default_factory=InsightSeoInputSchema,
    )


class InsightArticleUpdateSchema(
    InsightArticleCreateSchema
):
    change_summary: str = Field(
        default="",
        max_length=300,
    )


class InsightScheduleSchema(Schema):
    scheduled_for: datetime


class InsightRevisionSchema(Schema):
    id: UUID
    revision_number: int
    snapshot: dict[str, Any]
    change_summary: str
    created_at: datetime


class InsightArticleSchema(Schema):
    id: UUID
    title: str
    slug: str
    excerpt: str
    content: dict[str, Any]

    category_id: UUID | None
    category_name: str | None

    author_id: int | None
    author_email: str | None

    featured_image_id: UUID | None

    status: str
    published_at: datetime | None
    scheduled_for: datetime | None

    reading_time_minutes: int
    word_count: int
    view_count: int

    is_featured: bool
    is_active: bool
    allow_comments: bool
    is_publicly_available: bool

    current_revision_number: int

    tags: list[dict[str, Any]]
    related_articles: list[dict[str, Any]]
    internal_links: list[dict[str, Any]]

    seo: dict[str, Any] | None
    revisions: list[InsightRevisionSchema]
    publishing_events: list[dict[str, Any]]

    created_at: datetime
    updated_at: datetime

File: apps/dashboard/src/features/insights/schemas.ts
import { z } from "zod";

import { insightStatuses } from "./types";

export const insightCategorySchema = z.object({
  id: z.string(),
  name: z.string(),
  slug: z.string(),
  description: z.string(),
  parent_id: z.string().nullable(),
  is_active: z.boolean(),
  sort_order: z.number(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const insightTagSchema = z.object({
  id: z.string(),
  name: z.string(),
  slug: z.string(),
  description: z.string(),
  is_active: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
});

const insightArticleTagSchema = z.object({
  id: z.string(),
  name: z.string(),
  slug: z.string(),
});

const relatedInsightArticleSchema = z.object({
  id: z.string(),
  title: z.string(),
  slug: z.string(),
});

const insightInternalLinkSchema = z.object({
  id: z.string(),
  target_article_id: z.string(),
  target_article_title: z.string(),
  anchor_text: z.string(),
  context: z.string(),
  is_active: z.boolean(),
});

const insightSeoSchema = z.object({
  id: z.string(),
  meta_title: z.string(),
  meta_description: z.string(),
  canonical_url: z.string(),
  robots_index: z.boolean(),
  robots_follow: z.boolean(),
  open_graph_title: z.string(),
  open_graph_description: z.string(),
  open_graph_image_id: z.string().nullable(),
  twitter_title: z.string(),
  twitter_description: z.string(),
  article_schema: z.record(z.string(), z.unknown()),
  faq_schema: z.array(z.record(z.string(), z.unknown())),
});

const insightRevisionSchema = z.object({
  id: z.string(),
  revision_number: z.number(),
  snapshot: z.record(z.string(), z.unknown()),
  change_summary: z.string(),
  created_at: z.string(),
});

const insightPublishingEventSchema = z.object({
  id: z.string(),
  event_type: z.string(),
  description: z.string(),
  metadata: z.record(z.string(), z.unknown()),
  created_at: z.string(),
});

export const insightArticleSchema = z.object({
  id: z.string(),
  title: z.string(),
  slug: z.string(),
  excerpt: z.string(),
  content: z.record(z.string(), z.unknown()),
  category_id: z.string().nullable(),
  category_name: z.string().nullable(),
  author_id: z.number().nullable(),
  author_email: z.string().nullable(),
  featured_image_id: z.string().nullable(),
  status: z.enum(insightStatuses),
  published_at: z.string().nullable(),
  scheduled_for: z.string().nullable(),
  reading_time_minutes: z.number(),
  word_count: z.number(),
  view_count: z.number(),
  is_featured: z.boolean(),
  is_active: z.boolean(),
  allow_comments: z.boolean(),
  is_publicly_available: z.boolean(),
  current_revision_number: z.number(),
  tags: z.array(insightArticleTagSchema),
  related_articles: z.array(relatedInsightArticleSchema),
  internal_links: z.array(insightInternalLinkSchema),
  seo: insightSeoSchema.nullable(),
  revisions: z.array(insightRevisionSchema),
  publishing_events: z.array(insightPublishingEventSchema),
  created_at: z.string(),
  updated_at: z.string(),
});

export const paginatedInsightsSchema = z.object({
  items: z.array(insightArticleSchema),
  pagination: z.object({
    page: z.number(),
    page_size: z.number(),
    total_items: z.number(),
    total_pages: z.number(),
  }),
});

export const messageSchema = z.object({
  status: z.string(),
  message: z.string(),
});
